tools/comdb2ar/increment.cpp

```diff
@@ -0,0 +1,523 @@
+                    pages.clear();
+                    return true;
+                // otherwise just add the remaining pages
+                } else {
+                    file_expanded = true;
+                }
+                break;
+            }
+
+            // If a diff has been selected in a page, keep track of that page
+            if (assert_cksum_lsn(new_pagebuf, old_pagebuf, pagesize)) {
+                pages.push_back(PGNO(new_pagebuf));
+                *data_size += pagesize;
+                ret = true;
+            }
+
+            bytesleft -= pagesize;
+        }
+
+        if(new_pagebuf) free(new_pagebuf);
+    }
+    return ret;
+}
+
+ssize_t serialise_incr_file(
+    const FileInfo& file,
+    std::vector<uint32_t> pages,
+    std::string incr_path
+)
+// For a file with changed pages, go through each changed page and serialise it
+{
+    const std::string& filename = file.get_filename();
+    const std::string& filepath = file.get_filepath();
+
+    int flags;
+    bool skip_iomap = false;
+    std::ostringstream ss;
+
+    // Ensure large file support
+    assert(sizeof(off_t) == 8);
+
+    struct stat st;
+    if(stat(filepath.c_str(), &st) == -1) {
+        ss << "cannot stat file: " << std::strerror(errno);
+        throw SerialiseError(filename, ss.str());
+    }
+
+    size_t pagesize = file.get_pagesize();
+    if(pagesize == 0) {
+        pagesize = 4096;
+    }
+
+    std::ifstream ifs (filepath, std::ifstream::in | std::ifstream::binary);
+
+    char *char_pagebuf = NULL;
+    uint8_t *pagebuf = NULL;
+
+#if ! defined  ( _SUN_SOURCE ) && ! defined ( _HP_SOURCE )
+    posix_memalign((void**) &char_pagebuf, 512, pagesize);
+    posix_memalign((void**) &pagebuf, 512, pagesize);
+#else
+    char_pagebuf = (char*) memalign(512, pagesize);
+    pagebuf = (uint8_t*) memalign(512, pagesize);
+#endif
+
+    ssize_t total_read = 0;
+    ssize_t total_written = 0;
+
+    std::string incrFilename = incr_path + "/" + filename + ".incr";
+
+    std::ofstream incrFile(incrFilename, std::ofstream::out |
+                            std::ofstream::in | std::ofstream::binary);
+
+    for(std::vector<uint32_t>::iterator
+            it = pages.begin();
+            it != pages.end();
+            ++it){
+
+        ifs.seekg(pagesize * *it, ifs.beg);
+        ifs.read(&char_pagebuf[0], pagesize);
+        ssize_t bytes_read = ifs.gcount();
+        if(bytes_read < pagesize) {
+            std::ostringstream ss;
+            ss << "read error";
+            throw SerialiseError(filename, ss.str());
+        }
+
+        std::memcpy(pagebuf, char_pagebuf, pagesize);
+
+        uint32_t cksum;
+        bool cksum_verified;
+        verify_checksum(pagebuf, pagesize, false, false,
+                            &cksum_verified, &cksum);
+        if(!cksum_verified){
+            std::ostringstream ss;
+            ss << "serialise_file:page failed checksum verification";
+            throw SerialiseError(filename, ss.str());
```

Comment:
You should retry this.  Checksum failure can happen legitimately if we happen to read a page which was in the process of being paged-out.  Copy the logic in serialise_file.